fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/FDBDirectory.java

```diff
@@ -157,29 +159,56 @@ public FDBDirectory(@Nonnull Subspace subspace, @Nonnull FDBRecordContext contex
                 .maximumSize(maximumSize)
                 .recordStats()
                 .build();
+        this.fileSequenceCounter = new AtomicLong(-1);
         this.compressionEnabled = Objects.requireNonNullElse(context.getPropertyStorage().getPropertyValue(LuceneRecordContextProperties.LUCENE_INDEX_COMPRESSION_ENABLED), false);
         this.encryptionEnabled = Objects.requireNonNullElse(context.getPropertyStorage().getPropertyValue(LuceneRecordContextProperties.LUCENE_INDEX_ENCRYPTION_ENABLED), false);
         this.fileReferenceMapSupplier = Suppliers.memoize(this::loadFileReferenceCacheForMemoization);
     }
 
+    private long deserializeFileSequenceCounter(@Nullable byte[] value) {
+        return value == null ? 0L : Tuple.fromBytes(value).getLong(0);
+    }
+
+    @Nonnull
+    private byte[] serializeFileSequenceCounter(long value) {
+        return Tuple.from(value).pack();
+    }
+
+    @Nonnull
+    private CompletableFuture<Void> loadFileSequenceCounter() {
+        long originalValue = fileSequenceCounter.get();
+        if (originalValue >= 0) {
+            // Already loaded
+            return AsyncUtil.DONE;
+        }
```

Comment:
what if after one thread sees a negative counter above this line, but before it reads the value from storage below, another thread updates the counter, but has not mutated the value from storage yet, and this thread loads a stale value from storage?

Reply:
I'm having a bit of a hard time following your hypothetical. Are you saying something like this:

1. Thread 1: Gets negative value from fileSequenceCounter
2. Thread 2: Gets negative value from fileSequenceCounter
3. Thread 2: Loads value from storage
4. Thread 2: Puts loaded value into fileSequenceCounter
5. Thread 2: Updates counter value in fileSequenceCounter
6. Thread 1: Loads value from storage
7. Thread 2: Updates value in storage
8. Thread 1: Puts loaded value info fileSequenceCounter

Then, in this case, yes, at step 6, thread 1 will read a value from the database that does not include the update made at step 5 by thread 2. However, note that the `fileSequenceCounter` is updated by calling [`compareAndSet`](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/atomic/AtomicInteger.html#compareAndSet-int-int-) on line 188, and so it will only be updated if the `fileSequenceCounter` is still negative. So, in the case outlined above, step 8 will actually not update the `fileSequenceCounter`, and so even though thread 1 read a stale value from the database, it will actually get a value from `getIncrement` based on an up-to-date value, so I don't think there's a race

Reply:
yes that is what I meant. Sorry I just noticed that this stale value is to be CAS updated to the counter instead of being returned directly. Yeah then this looks good to me